src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs

```diff
@@ -207,21 +207,78 @@ public static unsafe void GetNativeVariantForObject(object? obj, IntPtr pDstNati
+                            break;
+                        case TypeCode.DBNull:
+                            data->SetAsNULL();
+                            break;
+                        case TypeCode.Boolean:
+                            data->AsBool = value.ToBoolean();
+                            break;
+                        case TypeCode.Char:
+                            data->AsUi2 = value.ToChar();
+                            break;
+                        case TypeCode.SByte:
+                            data->AsI1 = value.ToSByte();
+                            break;
+                        case TypeCode.Byte:
+                            data->AsUi1 = value.ToByte();
+                            break;
+                        case TypeCode.Int16:
+                            data->AsI2 = value.ToInt16();
+                            break;
+                        case TypeCode.UInt16:
+                            data->AsUi2 = value.ToUInt16();
+                            break;
+                        case TypeCode.Int32:
+                            data->AsI4 = value.ToInt32();
+                            break;
+                        case TypeCode.UInt32:
+                            data->AsUi4 = value.ToUInt32();
+                            break;
+                        case TypeCode.Int64:
+                            data->AsI8 = value.ToInt64();
+                            break;
+                        case TypeCode.UInt64:
+                            data->AsUi8 = value.ToUInt64();
+                            break;
+                        case TypeCode.Single:
+                            data->AsR4 = value.ToSingle();
+                            break;
+                        case TypeCode.Double:
+                            data->AsR8 = value.ToDouble();
+                            break;
+                        case TypeCode.Decimal:
+                            data->AsDecimal = value.ToDecimal();
+                            break;
+                        case TypeCode.DateTime:
+                            data->AsDate = value.ToDateTime();
+                            break;
+                        case TypeCode.String:
+                            data->AsBstr = value.ToString();
+                            break;
+                        default:
+                            throw new NotSupportedException();
+                    } 
+                    break;
+                case CriticalHandle:
+                    throw new ArgumentException();
+                case SafeHandle:
+                    throw new ArgumentException();
+                case Array:
+                    // SAFEARRAY implementation goes here.
+                    throw new NotSupportedException();
+                case ValueType:
+                    throw new NotSupportedException();
+                default:
+                    data->AsDispatch = obj;
```

Comment:
I suspect that you will pretty soon hit the case that needs to QueryInterface for IDispatch and fallback to IUnknown if the object does not implement IDispatch.

(I know that this is pre-existing issue.)

Reply:
For now simple WinForms working with this PR. Next thing would be `PreserveSig = false`. Until I play with WebBrowser, hopefully I would not care about IDispatch at all.